magenta/protobuf/music.proto

```diff
@@ -0,0 +1,125 @@
+message NoteSequence {
+  // NOTE: The first three fields are duplicated from SequenceFile to allow
+  // creation of portable SSTables of Sequence protos.
+  // Unique id. See data_utils.h GenerateSequenceId().
+  string id = 1;
+  // The path of the file relative to the root of the collection.
+  string filename = 2;
+  // The collection from which the file comes. This can be shorthand e.g.
+  // "bach". One purpose is to allow for easy selection of all or some files
+  // from a particular source.
+  string collection = 3;
+
+  // MIDI ticks per beat, also known as resolution or PPQ ("parts per quarter").
+  // There is no widely-used default. 220 is assumed per the choice made in
+  // third_party/py/pretty_midi.
+  int32 ticks_per_beat = 4;
+  // Lacking a time signature, 4/4 is assumed per MIDI standard.
+  repeated TimeSignature time_signatures = 5;
+  // Lacking a key signature, C Major is assumed per MIDI standard.
+  repeated KeySignature key_signatures = 6;
+  // Lacking a tempo change, 120 bpm is assumed per MIDI standard.
+  repeated TempoChange tempo_changes = 7;
+  // A Note combines a MIDI NoteOn and NoteOff into one event with duration.
+  repeated Note notes = 8;
+  // The total time of the Sequence in seconds.
+  double total_time = 9;
+
+  // MIDI-specific events that are generally relevant for performance and
+  // re-synthesis, but not for processing the music score.
+  repeated PitchBend pitch_bends = 10;
+  repeated ControlChange control_changes = 11;
+
+  message Note {
+    // MIDI pitch; see en.wikipedia.org/wiki/MIDI_Tuning_Standard for details.
+    int32 pitch = 1;
+    // Velocity ranging between 0 and 127.
+    int32 velocity = 2;
+    // Start and end time in seconds.
+    double start_time = 3;
+    double end_time = 4;
+    // Score-relative note length. E.g. a quarter note is 1/4.
+    int32 numerator = 5;
+    int32 denominator = 6;
+    // The pretty_midi instrument index, which in py_midi is the channel.
+    int32 instrument = 7;
+    // MIDI program is the sound playing (e.g. harpsichord).
+    int32 program = 8;
+  }
+
+  message TimeSignature {
+    // Time in seconds.
+    double time = 1;
+    int32 numerator = 2;
+    int32 denominator = 3;
+  }
+
+  message KeySignature {
+    // Time in seconds.
+    double time = 1;
+    Key key = 2;
+    Mode mode = 3;
+
+    enum Key {
+      option allow_alias = true;
+      C = 0;
+      C_SHARP = 1;
+      D_FLAT = 1;
+      D = 2;
+      D_SHARP = 3;
+      E_FLAT = 3;
+      E = 4;
+      F = 5;
+      F_SHARP = 6;
+      G_FLAT = 6;
+      G = 7;
+      G_SHARP = 8;
+      A_FLAT = 8;
+      A = 9;
+      A_SHARP = 10;
+      B_FLAT = 10;
+      B = 11;
+    }
+
+    enum Mode {
+      MAJOR = 0;
+      MINOR = 1;
+    }
+  }
+
+  message TempoChange {
+    // Time in seconds.
+    double time = 1;
+    // Tempo in beats per minute.
+    double bpm = 2;
+  }
+
+  // Stores MIDI PitchBend data. See the MIDI specfication for details.
```

Comment:
Can you link to the portion of the MIDI spec? There's a bunch of oddness in this message that'd be useful to look up.